### ISSUE TITLE
Can't install a gem from git from a specific commit

### ISSUE THREAD

**Issue by Nowaker** [bundler/bundler#4659]
This works:

``` ruby
gem 'middleman-hashicorp', git: 'https://github.com/hashicorp/middleman-hashicorp.git'
```

Result:

```
Using middleman-hashicorp 0.2.0 from https://github.com/hashicorp/middleman-hashicorp.git (at master@adc9159)
```

However, neither of these work:

``` ruby
gem 'middleman-hashicorp', git: 'https://github.com/hashicorp/middleman-hashicorp.git', ref: 'adc9159'
gem 'middleman-hashicorp', '0.2.0', git: 'https://github.com/hashicorp/middleman-hashicorp.git', ref: 'adc9159'
```

Result:

```
Could not find gem 'middleman-hashicorp' in https://github.com/hashicorp/middleman-hashicorp.git (at adc9159@adc9159).
Source contains 'middleman-hashicorp' at: 0.2.0 ruby
```

```
Could not find gem 'middleman-hashicorp (= 0.2.0)' in https://github.com/hashicorp/middleman-hashicorp.git (at adc9159@adc9159).
Source contains 'middleman-hashicorp' at: 0.2.0 ruby
```

I'm a little ashamed, as it feels like some stupid typo or something, but I was trying to make it for half an hour without a success, so I have no other conclusion than this being a bug in bundler.

**Comment by segiddins** [comment]
```
/Users/segiddins/development/opensource/bundler/lib/bundler/installer.rb:166:in `block in ensure_specs_are_compatible!': middleman-hashicorp-0.2.0 requires ruby version ~> 2.0.0, which is incompatible with the current version, ruby 2.3.1p112 (Bundler::InstallError)
```

**Comment by Nowaker** [comment]
Okay, it seems that I tested one of these without `ruby 'versionhere'` and that's why one of these succeeded and the other did not. However, the error message is super unclear, so I guess this ticket would be about improving the messaging.

**Comment by RochesterinNYC** [comment]
If that's the case, can we get more explicit reproduction steps (with specific `Gemfile`s, ruby versions, etc.)? Also, can you provide your bundler version via `bundle env` and try this on the latest version of bundler (`v.1.12.5`) if you're not currently on it?

**Comment by Nowaker** [comment]
Yes, sir, I'll prepare a test project that shows the error.

**Comment by LynnCo** [comment]
ping @Nowaker

**Comment by RochesterinNYC** [comment]
Hey @Nowaker, any update on this?

**Comment by njh** [comment]
I am having the same problem but by choosing a specific branch, rather than commit:

`$ bundle env`
```
Environment

    Bundler   1.12.5
    Rubygems  2.2.2
    Ruby      2.1.2p95 (2014-05-08 revision 45877) [x86_64-darwin14.0]
    Git       2.5.4 (Apple Git-61)

Bundler settings

    gem.coc
      Set for the current user (/Users/humfrn01/.bundle/config): "true"
    gem.mit
      Set for the current user (/Users/humfrn01/.bundle/config): "true"
    gem.test
      Set for the current user (/Users/humfrn01/.bundle/config): "minitest"
    orig_path
      Set via BUNDLE_ORIG_PATH: "/Users/humfrn01/.rbenv/versions/2.1.2/bin:/usr/local/Cellar/rbenv/1.0.0/libexec:/usr/local/heroku/bin:/Users/humfrn01/.rbenv/shims:/usr/local/opt/php56/bin:/Users/humfrn01/bin:/Users/humfrn01/pear/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"

Gemfile

    source "https://rubygems.org"
    ruby "2.1.2"
    
    gem 'json'
    gem 'rake'
    gem 'webrick'
    gem 'semverly'
    gem 'erubis'
    gem 'tilt'
    gem 'filesize'
    gem 'nokogiri'
    
    group :linkeddata do
      gem 'rdf'
      gem 'rdf-turtle'
      gem 'json-ld', :git => 'git@github.com:ruby-rdf/json-ld.git', :branch => 'develop'
    end
```

`$ bundle update`
```
Fetching git@github.com:ruby-rdf/json-ld.git
Fetching gem metadata from https://rubygems.org/
Fetching version metadata from https://rubygems.org/
Could not find gem 'json-ld' in git@github.com:ruby-rdf/json-ld.git (at
develop@d1a55df).
Source contains 'json-ld' at: 2.0.0.1 ruby
```

**Comment by njh** [comment]
Deleted my previous comment - my problem was that I was requiring a specific version of ruby in my Gemfile, which was in conflict with one of the gems using. Removing the ruby version requirement fixed me problem.

**Comment by Nowaker** [comment]
> Hey @Nowaker, any update on this?

Sorry - I forgot about it. Actually, it seems @njh provided everything you asked my for... so I'm repasting @njh's comment. His problem is exactly what I reported. When something doesn't work out in dependency resolution, a very weird error is shown that a typical human doesn't understand.

---

```
I am having the same problem but by choosing a specific branch, rather than commit:

$ bundle env

Environment

    Bundler   1.12.5
    Rubygems  2.2.2
    Ruby      2.1.2p95 (2014-05-08 revision 45877) [x86_64-darwin14.0]
    Git       2.5.4 (Apple Git-61)

Bundler settings

    gem.coc
      Set for the current user (/Users/humfrn01/.bundle/config): "true"
    gem.mit
      Set for the current user (/Users/humfrn01/.bundle/config): "true"
    gem.test
      Set for the current user (/Users/humfrn01/.bundle/config): "minitest"
    orig_path
      Set via BUNDLE_ORIG_PATH: "/Users/humfrn01/.rbenv/versions/2.1.2/bin:/usr/local/Cellar/rbenv/1.0.0/libexec:/usr/local/heroku/bin:/Users/humfrn01/.rbenv/shims:/usr/local/opt/php56/bin:/Users/humfrn01/bin:/Users/humfrn01/pear/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"

Gemfile

    source "https://rubygems.org"
    ruby "2.1.2"

    gem 'json'
    gem 'rake'
    gem 'webrick'
    gem 'semverly'
    gem 'erubis'
    gem 'tilt'
    gem 'filesize'
    gem 'nokogiri'

    group :linkeddata do
      gem 'rdf'
      gem 'rdf-turtle'
      gem 'json-ld', :git => 'git@github.com:ruby-rdf/json-ld.git', :branch => 'develop'
    end
$ bundle update

Fetching git@github.com:ruby-rdf/json-ld.git
Fetching gem metadata from https://rubygems.org/
Fetching version metadata from https://rubygems.org/
Could not find gem 'json-ld' in git@github.com:ruby-rdf/json-ld.git (at
develop@d1a55df).
Source contains 'json-ld' at: 2.0.0.1 ruby
```

**Comment by segiddins** [comment]
This should have a better error message now?

**Comment by segiddins** [comment]
As of Bundler 1.15.1, the error is:

```
Bundler could not find compatible versions for gem "ruby":
  In Gemfile:
    ruby (~> 2.1.2.0)

    json-ld was resolved to 2.1.4, which depends on
      ruby (>= 2.2.2)
```
